src/Searching.js

```diff
@@ -98,14 +146,318 @@ async function localSearch(searchTerm, roomId = undefined) {
 
     const localResult = await eventIndex.search(searchArgs);
 
+    if (processResult) {
+        emptyResult.seshatQuery.next_batch = localResult.next_batch;
+
+        const response = {
+            search_categories: {
+                room_events: localResult,
+            },
+        };
+
+        return MatrixClientPeg.get()._processRoomEventsSearch(emptyResult, response);
+    }
+
+    searchArgs.next_batch = localResult.next_batch;
+
+    const result = {
+        response: localResult,
+        query: searchArgs,
+    };
+
+    return result;
+}
+
+async function localPagination(searchResult) {
+    const eventIndex = EventIndexPeg.get();
+
+    const searchArgs = searchResult.seshatQuery;
+
+    const localResult = await eventIndex.search(searchArgs);
+    searchResult.seshatQuery.next_batch = localResult.next_batch;
+
     const response = {
         search_categories: {
             room_events: localResult,
         },
     };
 
-    const result = MatrixClientPeg.get()._processRoomEventsSearch(
-        emptyResult, response);
+    const result = MatrixClientPeg.get()._processRoomEventsSearch(searchResult, response);
+    searchResult.pendingRequest = null;
+
+    return result;
+}
+
+function compareOldestEvents(firstResults, secondResults) {
+    try {
+        const oldestFirstEvent = firstResults.results[firstResults.results.length - 1].result;
+        const oldestSecondEvent = secondResults.results[secondResults.results.length - 1].result;
+
+        if (oldestFirstEvent.origin_server_ts <= oldestSecondEvent.origin_server_ts) {
+            return -1;
+        } else {
+            return 1;
+        }
+    } catch {
+        return 0;
+    }
+}
+
+function combineEventSources(previousSearchResult, response, a, b) {
+    // Merge event sources and sort the events.
+    const combinedEvents = a.concat(b).sort(compareEvents);
+    // Put half of the events in the response, and cache the other half.
+    response.results = combinedEvents.slice(0, SEARCH_LIMIT);
+    previousSearchResult.cachedEvents = combinedEvents.slice(SEARCH_LIMIT);
+}
+
+/**
+ * Combine the events from our event sources into a sorted result
```

Comment:
Thanks for the detailed example, it makes it much easier to follow the approach! 😄

Reply:
It was a bit tricky to figure out if only moving one window was enough and I forgot my conclusion over the weekend so it's a helper for my aging brain as well :brain:.